### PR TITLE
Add Cron Jobs names to New Relic transactions

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
Feature: Add missing transaction names for Cron Jobs in New Relic reports
image

I've modified private methods, as this is not backward incompatible.

### Fixed Issues (if relevant)
- #22047
- #22061

### Manual testing scenarios (*)
1. Run `bin/magento cron:run`
2. Enjoy the transaction labels for single cron jobs in NewRelic

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by pmclain — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @pmclain, thank you for the review.
[ENGCOM-5488](https://jira.corp.magento.com/browse/ENGCOM-5488) has been created to process this Pull Request
:eight_spoked_asterisk: @pmclain, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by pmclain** [comment]
cc @mpchadwick

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php [hunk 1]
```diff
@@ -69,6 +71,11 @@ class ProcessCronQueueObserver implements ObserverInterface
      */
     const LOCK_PREFIX = 'CRON_GROUP_';
 
+    /**
+     * Cron Job name pattern for Profiling
+     */
+    const CRON_TIMERID = 'job %s';
```
Comment (sidolov) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: Please, make const private

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php [hunk 2]
```diff
@@ -351,40 +358,55 @@ protected function _runJob($scheduledTime, $currentTime, $jobConfig, $schedule,
...
     }
 
     /**
      * Stops profiling
      *
+     * @param string $jobName
      * @return void
      */
-    private function stopProfiling()
+    private function stopProfiling(string $jobName = '')
```
Comment (sidolov) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: I can't find the case when `$jobName` may be optional.

File: app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php [hunk 3]
```diff
@@ -351,40 +358,55 @@ protected function _runJob($scheduledTime, $currentTime, $jobConfig, $schedule,
...
     }
 
     /**
      * Starts profiling
      *
+     * @param string $jobName
      * @return void
      */
-    private function startProfiling()
+    private function startProfiling(string $jobName = '')
```
Comment (sidolov) [inline app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php]: I can't find the case when $jobName may be optional.

File: app/code/Magento/NewRelicReporting/Plugin/StatPlugin.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+use Magento\NewRelicReporting\Model\Config;
+use Magento\NewRelicReporting\Model\NewRelicWrapper;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Class StatPlugin handles single Cron Jobs transaction names
+ */
+class StatPlugin
+{
+    const TIMER_NAME_CRON_PREFIX = 'job';
```
Comment (sidolov) [inline app/code/Magento/NewRelicReporting/Plugin/StatPlugin.php]: Make const private

File: app/code/Magento/NewRelicReporting/Model/NewRelicWrapper.php [hunk 5]
```diff
@@ -67,6 +67,19 @@ public function setTransactionName(string $transactionName): void
         }
     }
 
+    /**
+     * Wrapper for 'newrelic_end_transaction'
+     *
+     * @param bool $ignore
+     * @return void
+     */
+    public function endTransaction($ignore = false)
```
Comment (sidolov) [inline app/code/Magento/NewRelicReporting/Model/NewRelicWrapper.php]: Unfortunately, we are not allowed to add new public methods according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

**Review by sidolov — CHANGES_REQUESTED** [review]
Please, see review comment and cover changes with automated tests

File: app/code/Magento/NewRelicReporting/Plugin/StatPlugin.php (hunk 4 above)
Comment (lbajsarowicz) [reply app/code/Magento/NewRelicReporting/Plugin/StatPlugin.php]: Does it mean that I can throw this PR to the trash as long as we don't have 2.4-dev branch?

File: app/code/Magento/NewRelicReporting/Model/NewRelicWrapper.php (hunk 5 above)
Comment (lbajsarowicz) [reply app/code/Magento/NewRelicReporting/Model/NewRelicWrapper.php]: Does it mean that I can throw this PR to the trash as long as we don't have 2.4-dev branch?

**Comment by lbajsarowicz** [comment]
Added ugly workaround, checking if any transaction is open - if so, close it with private method.

Everything should be removed with 2.4-dev

**Review by sidolov — APPROVED** [review]

**Comment by lbajsarowicz** [comment]
Turning down the issue to fix it differently, cleaner way.

**Pull request closed without merging**